### ISSUE TITLE
关于model.py中的DenoisingNet类的l0_norm函数

### ISSUE THREAD

**Issue by akajinchen** [HKUDS/AdaGCL#4]
def l0_norm(self, log_alpha, beta):
	gamma = args.gamma
	zeta = args.zeta
	gamma = torch.tensor(gamma)
	zeta = torch.tensor(zeta)
	reg_per_weight = torch.sigmoid(log_alpha - beta * torch.log(-gamma/zeta))这段代码的作用和原理我没有理解，希望作者可以给予解答

**Comment by yangqin-jiang** [comment]
Hi,

Thank you for your question.

The purpose of the l0_norm is to calculate the l0 regularization term for each weight in self.edge_weights, which represents the list of edge weights in the sparse matrix. The l0 regularization term corresponds to the number of non-zero elements. We aim to make the denoised graph as sparse as possible, which aligns with the objective of minimizing the number of non-zero elements in the optimized edge weights list. The definition of this loss is provided by Equation 8 in the paper. The l0_norm function implements the cumulative distribution function (CDF) for the edge weights. Since the edge weights have been extended and clipped with hyper parameters (i.e., gamma and zeta), the computation of this l0_norm becomes more complex.

If you have more questions, feel free to ask me.

Yangqin

**Comment by akajinchen** [comment]
这个去噪模块，好像最后并没有产生一个0，1组成的mask，而是生成了分布在[0,1]之间的系数？更像一个参数化的gat?

**Comment by yangqin-jiang** [comment]
对的，并没有产生离散的0，1组成的mask，而是生成了分布在[0, 1]之前的权重，GAT可能更关注节点的权重，我们的去噪模块主要是对图上的边的权重进行学习，让噪声边的权重尽可能得小，以达到去噪的效果。
